keps/sig-architecture/20191001-prevent-permabeta.md

```diff
@@ -0,0 +1,127 @@
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+APIs and features should not languish in beta.  They should take feedback and progress towards GA by either
+1. meeting GA criteria and getting promoted, or
+2. having a new beta and deprecating the previous beta
+  
+This must happen within six months (two releases).  If it does not,
+the API will be deprecated with an announced intent to remove the feature per the deprecation policy.
+
+## Motivation
+
+When a feature reaches beta, it is turned on by default.  This is great for getting feedback, but it can also lead to state
+where users and vendors start building important infrastructure against APIs that are not considered stable.
+In addition, once a feature is on by default, the incentive to further stabilize appears to diminish.
+See the features that have been beta for a long time: CSRs and Ingresses as examples.
+If we're honest with ourselves, a single actor has been cleaning up behind a lot of the project to unstick perma-beta APIs.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+1. Prevent APIs and features from being in a single beta version for more than six months.
+2. Prevent beta APIs from being treated as GA by users and vendors.
+
+### Non-Goals
+
+1. Promote APIs to GA before they are ready.
+
+## Proposal
+
+Once an API or feature reaches beta, it has six months to 
```

Comment:
A few examples here would be helpful (fenceposts are tricky with months... e.g. is six months 2 or 3 releases?)

For example, in v1.16, v1beta1 is released. Sample release note and API doc:
> * "The v1beta1 version of this API will be evaluated during v1.16 and v1.17, then deprecated in v1.18 (in favor of a new beta version, a GA version, or with no replacement), then removed in v1.21"

Scenario A - progression to v1beta2 in v1.18. Sample release note and API doc:
> * "The v1beta1 version of this API is deprecated in favor of v1beta2, and will be removed in v1.21"
> * "The v1beta2 version of this API will be evaluated during v1.18 and v1.19, then deprecated in v1.20 (in favor of a new beta version, a GA version, or with no replacement), then removed in v1.23"

Scenario B - progression to v1 in v1.18. Sample release note and API doc:
> * "The v1beta1 version of this API is deprecated in favor of v1, and will be removed in v1.21"

Scenario C - deprecation with no replacement. Sample release note and API doc:
> * "The v1beta1 version of this API is deprecated with no replacement, and will be removed in v1.21"